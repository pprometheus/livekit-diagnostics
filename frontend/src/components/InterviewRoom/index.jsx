import React, { useState, useEffect } from "react";
import TopBar from "../TopBar";
import Controls from "../Controls";
import Sidebar from "../SideBar";
import { RoomAudioRenderer, RoomContext } from "@livekit/components-react";
import "@livekit/components-styles";
import Chart from "../Chart";
import { MyVideoConference } from "../../container/InterviewRoomContainer/interviewRoomContainer";

const InterviewRoom = ({ token, roomA, statsData}) => {
  const [isSidebarVisible, setIsSidebarVisible] = useState(true);

  const toggleSidebar = () => setIsSidebarVisible((v) => !v);


  useEffect(() => {
    const handleKeyPress = (e) => {
      if (e.shiftKey && e.key === "D") {
        setIsSidebarVisible((prev) => !prev);
      }
    };
    window.addEventListener("keydown", handleKeyPress);
    return () => window.removeEventListener("keydown", handleKeyPress);
  }, []);

  if (!token) return <div>Loading...</div>;

  return (
    <div className="w-full h-screen bg-gray-100 dark:bg-gray-900 flex flex-col">
      <RoomContext.Provider value={roomA}>
        <TopBar />

        <div className="flex flex-1 overflow-hidden">
          <MyVideoConference />
          <RoomAudioRenderer />

          {isSidebarVisible && (
            <Sidebar>
              <span className="text-sm text-white">Network Connected via:</span>
              <Chart
                title="Bandwidth (Mbps)"
                data={statsData}
                xDataKey="time"
                yDomain={[0, 100]}
                yTickFormatter={(value) => `${(value / 1e6).toFixed(2)}`}
                lines={[
                  {
                    dataKey: "download",
                    name: "Download (bytes)",
                    stroke: "#00bfff",
                  },
                  {
                    dataKey: "upload",
                    name: "Upload (bytes)",
                    stroke: "#cc88ff",
                  },
                ]}
              />

              <Chart
                title="Latency (ms)"
                data={statsData}
                xDataKey="time"
                yDomain={[0, 50]}
                lines={[
                  { dataKey: "latency", name: "RTT (ms)", stroke: "#ff7300" },
                ]}
              />

              <Chart
                title="Packet Loss (%)"
                data={statsData}
                xDataKey="time"
                yDomain={[0, 20]}
                lines={[
                  {
                    dataKey: "lossFraction",
                    name: "Loss (%)",
                    stroke: "#ff0000",
                  },
                ]}
              />
            </Sidebar>
          )}
        </div>

        <Controls onToggleSidebar={toggleSidebar}/>
      </RoomContext.Provider>
    </div>
  );
};

export default InterviewRoom;
