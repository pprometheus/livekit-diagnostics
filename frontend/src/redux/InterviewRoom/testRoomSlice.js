import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  peers: [],
};

const testRoomSlice = createSlice({
  name: "testRoom",
  initialState,
  reducers: {
    fetchToken: (state, action) => {
      const peerName = action.payload;
      console.log("peerName",peerName)
      const existingPeer = state.peers.find((p) => p.name === peerName);

      if (existingPeer) {
        existingPeer.status = "loading";
      } else {
        state.peers.push({
          name: peerName,
          token: null,
          roomName: null,
          status: "loading",
        });
      }
    },

    setToken: (state, action) => {
      const { peer, token, roomName } = action.payload;
      const existingPeer = state.peers.find((p) => p.name === peer);

      if (existingPeer) {
        existingPeer.token = token;
        existingPeer.roomName = roomName;
        existingPeer.status = "success";
      } else {
        state.peers.push({
          name: peer,
          token,
          roomName,
          status: "success",
        });
      }
    },

    tokenStatus: (state, action) => {
      const { peer, status } = action.payload;
      const existingPeer = state.peers.find((p) => p.name === peer);

      if (existingPeer) {
        existingPeer.status = status;
      }
    },
  },
});

export const { fetchToken, setToken, tokenStatus } = testRoomSlice.actions;

export const selectPeerByName = (name) => (state) =>
  state.testRoom.peers.find((p) => p.name === name);

export default testRoomSlice.reducer;
