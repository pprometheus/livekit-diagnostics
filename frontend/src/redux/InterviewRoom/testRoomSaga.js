import { takeLatest, put, call } from "redux-saga/effects";
import { fetchToken, fetchTokenB, setToken, tokenStatus } from "./testRoomSlice";
import {getToken} from "../../utils/helper";
function* handleToken(action) {
  const peer  = action.payload.participantName;
  console.log("Fetching:", peer, action.payload.participantName);
  try {
    const tokenData = yield call(getToken,peer); 
    yield put(setToken({ peer, token: tokenData, roomName:"Testing Room" }));
  } catch (error) {
    yield put(tokenStatus({ peer, status: "error" })); 
    console.error(`Error fetching token for ${peer}:`, error.message);
  }
}

function* testRoomSaga() {
  yield takeLatest(fetchToken.type, handleToken); 
  yield takeLatest(fetchTokenB.type, handleToken); 

}

export default testRoomSaga;
