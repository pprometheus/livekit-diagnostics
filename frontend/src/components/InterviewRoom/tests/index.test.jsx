import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import InterviewRoom from "..";
jest.mock("@livekit/components-styles", () => ({}));

jest.mock("../../TopBar", () => () => <div data-testid="topbar" />);
jest.mock("../../Controls", () => ({ onToggleSidebar }) => (
  <button data-testid="controls" onClick={onToggleSidebar} />
));
jest.mock("../../SideBar", () => ({ children }) => (
  <div data-testid="sidebar">{children}</div>
));
jest.mock("../../Chart", () => () => <div data-testid="chart" />);
jest.mock(
  "../../../container/InterviewRoomContainer/interviewRoomContainer",
  () => ({
    MyVideoConference: () => <div data-testid="videoconference" />,
  })
);

jest.mock("@livekit/components-react", () => ({
  RoomAudioRenderer: () => <div data-testid="audio-renderer" />,
  RoomContext: { Provider: ({ children }) => children },
}));

describe("InterviewRoom Component", () => {
  const mockToken = "test-token";
  const mockStatsData = [
    {
      time: "10:00 AM",
      download: 5000000,
      upload: 1000000,
      latency: 25,
      lossFraction: 0.5,
    },
  ];
  const mockRoomA = {};

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders main components when token exists", () => {
    render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );

    expect(screen.getByTestId("topbar")).toBeInTheDocument();
    expect(screen.getByTestId("videoconference")).toBeInTheDocument();
    expect(screen.getByTestId("audio-renderer")).toBeInTheDocument();
    expect(screen.getByTestId("controls")).toBeInTheDocument();
  });

  test("shows sidebar by default with charts", () => {
    render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );

    const sidebar = screen.getByTestId("sidebar");
    expect(sidebar).toBeInTheDocument();
    expect(screen.getAllByTestId("chart")).toHaveLength(3);
  });

  test("toggles sidebar via controls button", () => {
    render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );

    const button = screen.getByTestId("controls");
    fireEvent.click(button);
    expect(screen.queryByTestId("sidebar")).not.toBeInTheDocument();

    fireEvent.click(button);
    expect(screen.getByTestId("sidebar")).toBeInTheDocument();
  });

  test("toggles sidebar via Shift+D keyboard shortcut", () => {
    render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );

    expect(screen.getByTestId("sidebar")).toBeInTheDocument();

    fireEvent.keyDown(window, { key: "D", shiftKey: true });
    expect(screen.queryByTestId("sidebar")).not.toBeInTheDocument();

    fireEvent.keyDown(window, { key: "D", shiftKey: true });
    expect(screen.getByTestId("sidebar")).toBeInTheDocument();
  });

  test("passes correct props to Charts", () => {
    const { container } = render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );

    const charts = container.querySelectorAll('[data-testid="chart"]');
    expect(charts.length).toBe(3);
  });

  test("to match snapshot", () => {
    const { container } = render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("cleans up event listeners on unmount", () => {
    const { unmount } = render(
      <InterviewRoom
        token={mockToken}
        roomA={mockRoomA}
        statsData={mockStatsData}
      />
    );

    const originalRemoveListener = window.removeEventListener;
    window.removeEventListener = jest.fn();

    unmount();

    expect(window.removeEventListener).toHaveBeenCalledWith(
      "keydown",
      expect.any(Function)
    );

    window.removeEventListener = originalRemoveListener;
  });
});
