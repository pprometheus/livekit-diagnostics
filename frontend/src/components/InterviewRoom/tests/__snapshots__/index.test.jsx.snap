// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterviewRoom Component to match snapshot 1`] = `
<div>
  <div
    class="w-full h-screen bg-gray-100 dark:bg-gray-900 flex flex-col"
  >
    <div
      data-testid="topbar"
    />
    <div
      class="flex flex-1 overflow-hidden"
    >
      <div
        data-testid="videoconference"
      />
      <div
        data-testid="audio-renderer"
      />
      <div
        data-testid="sidebar"
      >
        <span
          class="text-sm text-white"
        >
          Network Connected via:
        </span>
        <div
          data-testid="chart"
        />
        <div
          data-testid="chart"
        />
        <div
          data-testid="chart"
        />
      </div>
    </div>
    <button
      data-testid="controls"
    />
  </div>
</div>
`;
